### PR TITLE
Generate server methods based on `lsprotocol` type definitions

### LINKED ISSUE #306: Explore parity between `server.LanguageServer` and `protocol.LanguageServerProtocol` methods
In `server.LanguageServer` we have a lot of methods that look like:
```python
def apply_edit(
        self, edit: WorkspaceEdit, label: Optional[str] = None
    ) -> WorkspaceApplyEditResponse:
        """Sends apply edit request to the client."""
        return self.lsp.apply_edit(edit, label)
```
Therefore, they do nothing but call out to the underlying `protocol.LanguageServerProtocol` method. There may be a good reason for this, but it isn't documented. There are some cases such as with `publish_diagnostics`, where there is an advantage because the `server.LanguageServer` method signature is simpler.

Let's either document the reasoning for the apparent repetition, or deprecate the duplicated server versions in favour of the protocol versions.

### PR DESCRIPTION
**Pull request by alcarney**
#487 introduced some breaking changes, so this PR is taking the opportunity to make a few more that were going to be part of #418.

Similar to the `BaseLanguageClient`, this PR updates the code generation script to now also produce a `BaseLanguageServer` with all the server side methods generated based on the type definitions provided by `lsprotocol`

Not only does this improve consistency and remove the need for the duplicated methods on the `LanguageServerProtocol` class (Closes #306), we also get all the server side methods we were missing!

Aside from the obvious naming/argument changes, this also introduces the following changes (which we will need to document in a migration guide)

- `server.progress` now sends a `$/progress` notification, rather than giving access to pygls' [`Progress`](https://github.com/openlawlibrary/pygls/blob/0f6005cf522f48889d7eb69e7d76c70ce75c119d/pygls/progress.py#L18) helper. The helper is now accessed via `server.work_done_progress`
- The underlying protocol class is now accessed via `server.protocol` rather than `server.lsp`.

Finally, the base `Server` class has been tidied up a little
- Removed some LSP specific quriks (e.g. `sync_kind`)
- The `report_server_error` code has been pushed down into the base
- The `feature`, `thread` and `command` registration code has been pushed down into the base

**TODO**

- [x] Merge #487 
- [x] Fix tests and example servers
- [x] Identify and document all breaking changes

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly

## Automated linters

You can run the lints that are run on CI locally with:
```sh
poetry install --all-extras --with dev
poetry run poe lint
```

[commit messages]: https://conventionalcommits.org/

### CODE REVIEW

**Comment by alcarney** [comment]
Added an initial draft of the migration guide, there are a few more changes we can make (e.g. removing deprecated methods) so I expect it will take a few more revisions to get right.

So this should be ready for a review now.

That said, the lint job has started [failing](https://github.com/openlawlibrary/pygls/actions/runs/10478488217/job/29022141986#step:7:14) and I'm not entirely sure why...

```
Run source $VENV
/home/runner/work/_temp/0f22c65c-8345-4c8d-8709-e3f4d68c7c6b.sh: /home/runner/work/pygls/pygls/.venv/bin/poe: /home/runner/work/pygls/pygls/.venv/bin/python: bad interpreter: No such file or directory
Error: Process completed with exit code 126.
```

File: scripts/check_client_is_uptodate.py [hunk 1]
```diff
@@ -1,20 +1,30 @@
...
-    print("✅ Pygls client is up to date")
+    print("✅ Pygls client and server are up to date")
 else:
     print(
-        "🔴 Pygls client not uptodate\n"
-        "1. Generate with: `poetry run poe generate_client`\n"
-        "2. Commit"
+        (
+            "🔴 Pygls client or server not up to date\n"
+            "1. Re-generate with: `poetry run poe generate_client`\n"
```
Comment (tombh) [inline scripts/check_client_is_uptodate.py:26]: It's `generate_client_server` now right?

File: pyproject.toml [hunk 2]
```diff
@@ -58,7 +58,7 @@ ruff = "ruff check ."
 mypy = "mypy -p pygls"
 check_generated_client = "python scripts/check_client_is_uptodate.py"
 check_commit_style = "npx commitlint --from origin/main --to HEAD --verbose --config commitlintrc.yaml"
-generate_client = "python scripts/generate_client.py --output pygls/lsp/client.py"
+generate_client = "python scripts/generate_client_server.py pygls/lsp"
```
Comment (tombh) [inline pyproject.toml:61]: Should we rename the command to `generate_client_server`?

File: scripts/check_client_is_uptodate.py
Comment (tombh) [inline scripts/check_client_is_uptodate.py:1]: Should we rename this to `check_generated_code_is_uptodate.py`?

**Review by tombh — CHANGES_REQUESTED** [review]
Such great work again ❤️ So many improvements, detailed docs and fixing old issues 💪

Do the new docs get automatically published on merge to main? I can't remember now. It might be good to have an update in our main README about how we're looking for beta testers for the breaking changes and a link to the migration docs?

**Comment by tombh** [comment]
I have a PR for the failing CI issue: https://github.com/openlawlibrary/pygls/pull/491

File: scripts/check_client_is_uptodate.py
Comment (alcarney) [reply scripts/check_client_is_uptodate.py:1]: Done

File: pyproject.toml (hunk 2 above)
Comment (alcarney) [reply pyproject.toml]: I've renamed it and the script to `generate_code` - seems to complement the `check_generated_code_is_uptodate.py` script?

**Comment by alcarney** [comment]
> Do the new docs get automatically published on merge to main?

Yes, https://pygls.readthedocs.io/en/latest/ points to `main`

>  It might be good to have an update in our main README about how we're looking for beta testers for the breaking changes and a link to the migration docs?

I've added a note to the README, though there are a few more changes we should make before shouting too loudly about it :)

**Review by tombh — APPROVED** [review]
Nice! LGTM 🚀

**Pull request merged**